### PR TITLE
Race Ban Code Improvement [Not Modular... Yet?]

### PR DESCRIPTION
**Pull request by SpaceDragon00**
## About The Pull Request

I really wish these problems could have been addressed in the initial PR, but it was already merged. If a PR is gonna be treated like a meme during the review phase (Where problems are replied to with "Dont care") then it should be treated as a meme during the merge phase.

This frustrated me enough to make me want to fix it at midnight

## Why It's Good For The Game

Bans are no longer evaluated every time spec_life is called, fewer lines are modified in tg files, might be able to fully modularize it

image

## Changelog
:cl:
admin: More races can be banned, and it's easier to add even more (Ban for EVERY race????)
code: Added a global list of Fulp bans in the format of the job ban list in the sql ban code to more efficiently modularize ban code
code: Reduced the number of lines overridden in a TG file from 5 to 1... for now
code: Moved the ban evaluation code off of spec_life and onto on_species_gain
code: Added a beefman species id define (Sorry pucci, I needed it early)
/:cl:

Elephant in the Room: Fully Modularizable?

To be frank I dont know if this is fully modularizable. Given the entirety of the ban panel proc takes place within a SINGLE proc, wrapped in opening and closing form tags, Im not sure if it can be modified before or after its runtime. You can receive information from a form, but modifying it is the tough question. Im not adept enough with code or HTML to know if you can delete the concluding tags, add the content you need, then re-add them or what

Perhaps someone more familiar with the formatting of a browser can use get_content() to retroactively remove the closing tags, before copy-pasting the code from long_job_lists in a modular folder to add the content before closing it once again. I might try this later but dont get your hopes up.

Looking at the code, why the fuck does get content not directly provide you with the content variable, making it infinitely harder to modify the existing content

### CODE REVIEW

**Comment by SpaceDragon00** [comment]
Marked as draft because I dont know if the SQL shit works yet-- The revert code seems to work fine roundstart and for manual species settings, but there may be some edge cases one has to worry about.

I removed the beefman meme code because of jank beef code and also possibly unintended interactions with bluespace prophet

**Comment by JohnFulpWillard** [comment]
What you're doing with bloodsucker/MH is terrible and will annoy the fuck out of admins trying to antag ban people.

**Comment by JohnFulpWillard** [comment]
> I removed the beefman meme code

That was not a meme.

**Comment by JohnFulpWillard** [comment]
Also as long as there's 1 edited line in a TG file, it's not modular, and having more will not change anything.

**Comment by SpaceDragon00** [comment]
Just so you know, clicking the "Antagonist Positions" category to select all, also selects the fulpstation antagonist positions on my current build (See Below)

The Good News: I have modularized bans

The Bad News: It is horribly fucking cursed and Im literally not going to commit it until I get permission from someone smarter than me

image


Proof it "works":

image

**Comment by Fikou** [comment]
@JohnFulpWillard dumby

File: fulp_modules/Z_edits/race_bans/races.dm [hunk 1]
```diff
@@ -1,15 +1,10 @@
-/datum/species/ethereal/spec_life(mob/living/carbon/human/H)
-	if(is_banned_from(H.ckey, RACE_ETHEREAL))
-		if(prob(50))
-			H.set_species(/datum/species/beefman)
-		else
-			H.set_species(/datum/species/human)
+/datum/species/on_species_gain(mob/living/carbon/C, datum/species/old_species, pref_load)
+	if(is_banned_from(C.ckey, src.id))
```
Comment (Fikou) [inline fulp_modules/Z_edits/race_bans/races.dm:2]: dont do src.var unless you have an arg named the same as a var on that object

File: fulp_modules/Z_edits/race_bans/races.dm [hunk 2]
```diff
@@ -1,15 +1,10 @@
...
-	if(is_banned_from(H.ckey, RACE_FELINID))
-		if(prob(50))
-			H.set_species(/datum/species/beefman)
-		else
-			H.set_species(/datum/species/human)
-	. = ..()
+/// Made into an individual proc to ensure that the to_chat message would always show to users. Sometimes it would not appear during roundstart as it would be sent too soon.
+/mob/living/carbon/proc/banned_species_revert()
+	to_chat(src, span_alert("You are currently banned from playing this race. Please review any ban messages you have received, and contact admins if you believe this is a mistake."))
+	src.set_species(/datum/species/human)
```
Comment (Fikou) [inline fulp_modules/Z_edits/race_bans/races.dm:10]: dont call procs on src

File: fulp_modules/Z_edits/race_bans/races.dm (hunk 1 above)
Comment (SpaceDragon00) [reply fulp_modules/Z_edits/race_bans/races.dm:2]: Noted for the future, was just being verbose. I'll fix it on my next go around

**Comment by Helianthus00** [comment]
Is there no way to make the races capitalized on the ban page? That'll definitely bother me forever. 

Also I don't love the lack of noting from the modular bans, but these race bans will be so rarely used that I think it's fine if that means it's fully modular.

**Comment by SpaceDragon00** [comment]
> 
> 
> Is there no way to make the races capitalized on the ban page? That'll definitely bother me forever.
> 
> Also I don't love the lack of noting from the modular bans, but these race bans will be so rarely used that I think it's fine if that means it's fully modular.

The code posted was left simple in order to more show what i was trying and people could tell me to not do that if it was horrendously stupid. My comment passes are made towards the end so that I fully understand the internal logic of what I'm trying to do without relying on them

I'll try and see if I can intercept the ID string and capitalize it but Im not sure how that would interact with SQL. I'll also see if there are alternative defines that exist already, and if not, I'll just bite the bullet and make more defines

**Comment by SpaceDragon00** [comment]
So I'm going to give Tom a matrix style option because both options suck for them

I tried to dynamically build a list of race names via proc overrides on things like /proc/make_datum_references_lists() and such but the code doesnt work. If you see any obvious errors check below, where i overrode a different proc. The .copy() is there out of desperation, and direct assignments (Without using "Output") didn't work either

image

So, option 1 is I hard code the name of every race into the list, which sucks because hard coding also sucks and we'll need to update the list manually if (For whatever reason) new races are added or the names of the races change

Option 2 is to accept that the text will be lowercase

The option that will make me cry is to personally modify the IDs every time they are accessed specifically to put the first letter in uppercase (Which I dont even know if there's an existing or DM proc that can do that)

Choose wisely

**Comment by Helianthus00** [comment]
Lowercase is fine if getting them uppercase is too much of a hassle. I only mentioned it in hopes that it'd be an easy fix or was an oversight or something.

**Comment by SpaceDragon00** [comment]
In that case I'll do some final testing and if no dire errors are found I'll set this as ready

**Comment by SpaceDragon00** [comment]
Going to bring up our disagreement here since, because John and I inherently disagree on this issue, we'll need a maintainer like @Helianthus00 to resolve it

Q: Should racebanned people have the possibility of being turned into a beefman

In my opinion, no. Giving people the ability to transmute into a beefman OR human at will by repeatedly turning into a banned race (Via mutation toxin, pride mirrors, or otherwise) will give banned people an advantage over others who are not, as they become able to access bluespace prophet by becoming a beefman through transmuting into their banned race, before turning back into a human by doing so again. This gives them the primary advantage of beefmen without factoring in the downsides after they turn back into a human, on top of giving them two appearances they can switch between at will, which is more than a normal mutation toxin would give you. Assuming that set_species doesn't additionally regenerate limbs, they can regenerate human limbs with this mechanic aswell.

As rare as these situations may be, I don't want there to be the possibility of powergaming bans (Especially self-requested ones). Im already hesitant of this system to begin with as Im not entirely sure how this functions with antagonists such as Changelings, plasmaman mutation toxins as an assassination device, banned plasmamen keeping plasmamen gear, or otherwise; though all of these concerns are seconday to the one mentioned previously.

John believes that the 50% beefman chance from before the refactor should be kept. As for the reason, you'll have to ask them. It would be unfair to represent his argument personally as someone opposed to their proposal.

**Comment by Helianthus00** [comment]
> Q: Should racebanned people have the possibility of being turned into a beefman

I liked the beefman thing as a meme and it's funny to see meatmen walking around uwuing and stuff but yes that doesn't have a place in a serious pr, let alone something that could possibly be powergamed. We can always setspecies to beefman anyway. Keeping it as a return to human only is fine.

**Pull request merged**